src/osd/OSD.h

```diff
@@ -540,44 +540,54 @@ class OSDService {
   }
 
   void activate_map();
-  /// map epochs reserved below
-  map<epoch_t, unsigned> map_reservations;
+  // the advanced seq
+  std::atomic_uint map_reserve_seq;
+  // the completed seq number
+  std::atomic_uint map_completed_seq;
+  std::atomic_uint map_reserved_waiters;
+  std::atomic_uint map_wait_version;
+  // wait_version -> pair(seq, count)
+  std::map<uint64_t, std::pair<uint64_t, uint64_t> > map_wait_pending;
 
   /// gets ref to next_osdmap and registers the epoch as reserved
-  OSDMapRef get_nextmap_reserved() {
-    Mutex::Locker l(pre_publish_lock);
-    if (!next_osdmap)
-      return OSDMapRef();
-    epoch_t e = next_osdmap->get_epoch();
-    map<epoch_t, unsigned>::iterator i =
-      map_reservations.insert(make_pair(e, 0)).first;
-    i->second++;
+  OSDMapRef get_nextmap_reserved(uint64_t &s) {
+    s = ++map_reserve_seq;
     return next_osdmap;
```

Comment:
http://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

I *think* you need to ensure that all accesses to the next_osdmap shared_ptr are done through the atomic accessors.  I don't think the barrier from map_reserve_seq is sufficient.